ext/opentelemetry-exporter-cloud-trace/src/opentelemetry/exporter/cloud_trace/__init__.py

```diff
@@ -0,0 +1,332 @@
+                    "span_id": span_id,
+                    "display_name": _get_truncatable_str_object(
+                        span.name, 128
+                    ),
+                    "start_time": start_time,
+                    "end_time": end_time,
+                    "parent_span_id": parent_id,
+                    "attributes": _extract_attributes(
+                        span.attributes, MAX_SPAN_ATTRS
+                    ),
+                    "links": _extract_links(span.links),
+                    "status": _extract_status(span.status),
+                    "time_events": _extract_events(span.events),
+                }
+            )
+            # TODO: Leverage more of the Cloud Trace API, e.g.
+            #  same_process_as_parent_span and child_span_count
+
+        return cloud_trace_spans
+
+    def shutdown(self):
+        pass
+
+
+def _get_hexadecimal_trace_id(trace_id: int) -> str:
+    return "{:032x}".format(trace_id)
+
+
+def _get_hexadecimal_span_id(span_id: int) -> str:
+    return "{:016x}".format(span_id)
+
+
+def _get_time_from_ns(nanoseconds: int) -> Dict:
+    """Given epoch nanoseconds, split into epoch milliseconds and remaining
+    nanoseconds"""
+    if not nanoseconds:
+        return None
+    seconds, nanos = divmod(nanoseconds, 1e9)
+    return {"seconds": int(seconds), "nanos": int(nanos)}
+
+
+def _get_truncatable_str_object(str_to_convert: str, max_length: int):
+    """Truncate the string if it exceeds the length limit and record the
+    truncated bytes count."""
+    truncated, truncated_byte_count = _truncate_str(str_to_convert, max_length)
+
+    return TruncatableString(
+        value=truncated, truncated_byte_count=truncated_byte_count
+    )
+
+
+def _truncate_str(str_to_check: str, limit: int) -> Tuple[str, int]:
+    """Check the length of a string. If exceeds limit, then truncate it."""
+    encoded = str_to_check.encode("utf-8")
+    truncated_str = encoded[:limit].decode("utf-8", errors="ignore")
+    return truncated_str, len(encoded) - len(truncated_str.encode("utf-8"))
+
+
+def _extract_status(status: trace_api.Status) -> Status:
+    """Convert a Status object to protobuf object."""
+    if not status:
+        return None
+    status_dict = {"details": None, "code": status.canonical_code.value}
+
+    if status.description is not None:
+        status_dict["message"] = status.description
+
+    return Status(**status_dict)
+
+
+def _extract_links(links: Sequence[trace_api.Link]) -> ProtoSpan.Links:
+    """Convert span.links"""
+    if not links:
+        return None
+    extracted_links = []
+    dropped_links = 0
+    if len(links) > MAX_NUM_LINKS:
+        logger.warning(
+            "Exporting more then %s links, some will be truncated",
+            MAX_NUM_LINKS,
+        )
+        dropped_links = len(links) - MAX_NUM_LINKS
+        links = links[:MAX_NUM_LINKS]
+    for link in links:
+        if len(link.attributes) > MAX_LINK_ATTRS:
```

Comment:
this chunk of code (along with the warning) could be refactored into _extract_attributes.

I see some boilerplate here and extract_events that includes the max check then logs a warning. This feels intuitive to me to put the warning around some attributes will be dropped into the _extract_attributes itself, or maybe even the BoundedDict itself.

Reply:
Hey Yusuke, so I also noticed that the max length check + warning is repeated in extract_events and extract_links. The reason I kept it that way is so I can print more informative warnings then just "Too many attributes, dropped some", doing it this way I can say which event or link had too many attributes.

Reply:
sounds good, although it feels like that's still achievable after a refactor.